### PR TITLE
clients/horizonclient: Add support for Accounts endpoint.

### PR DESCRIPTION
**Pull request by abuiles**
<details>
  <summary>PR Checklist</summary>

### PR Structure

* [ ] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [ ] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [ ] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [ ] This PR adds tests for the most critical parts of the new functionality or fixes.
* [ ] I've updated any docs ([developer docs](https://www.stellar.org/developers/reference/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [ ] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [ ] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### What

Extend the horizonclient to support accounts filtering. It allows
account retrival by signer or with a trustline to an asset.

```go
	client := horizonclient.DefaultPublicNetClient
        accountID := "GCLWGQPMKXQSPF776IU33AH4PZNOOWNAWGGKVTBQMIC5IMKUNP3E6NVU"
	accountsRequest := horizonclient.AccountsRequest{Signer: accountID}

	account, err := client.Accounts(accountsRequest)
	if err != nil {
		fmt.Println(err)
		return
	}

	fmt.Print(account)
```

### Why

This is part of the work in #2014, which is required to have full compatibility with Horizon 1.0.

### Known limitations

[TODO or N/A]

### CODE REVIEW

File: clients/horizonclient/client.go [hunk 1]
```diff
@@ -226,6 +226,17 @@ func (c *Client) HorizonTimeOut() time.Duration {
 	return c.horizonTimeOut
 }
 
+// Accounts returns accounts who have a given signer or
+// have a trustline to an asset.
+// See https://www.stellar.org/developers/horizon/reference/endpoints/accounts.html
+func (c *Client) Accounts(request AccountsRequest) (accounts hProtocol.AccountsPage, err error) {
+	err = c.sendRequest(request, &accounts)
+	if err != nil {
```
Comment (ire-and-curses) [inline clients/horizonclient/client.go]: Seems like this `if` is redundant - we return `err` either way

File: clients/horizonclient/main.go [hunk 2]
```diff
@@ -212,7 +213,18 @@ type HorizonRequest interface {
 	BuildURL() (string, error)
 }
 
-// AccountRequest struct contains data for making requests to the accounts endpoint of a horizon server.
+// AccountsRequest struct contains data for making requests to the accounts endpoint of a horizon server.
+// "Signer" or "Asset" fields should be set when retrieving AccountsData.
```
Comment (ire-and-curses) [inline clients/horizonclient/main.go]:
```suggestion
// Either "Signer" or "Asset" fields should be set when retrieving AccountsData.
```

File: clients/horizonclient/main.go [hunk 3]
```diff
@@ -212,7 +213,18 @@ type HorizonRequest interface {
...
+// At the moment, you can't use both filters are the same time.
+type AccountsRequest struct {
+	Signer string
+	Asset  string
+	Order  Order
+	Cursor string
+	Limit  uint
+}
+
+// AccountRequest struct contains data for making requests to the show account endpoint of a horizon server.
```
Comment (ire-and-curses) [inline clients/horizonclient/main.go]: It doesn't seem great that we have two structs with very similar names `AccountsRequest` vs `AccountRequest`. Is there something we can do to name them more clearly?

Comment (ire-and-curses) [reply clients/horizonclient/main.go]: (I don't really like the plural name AccountsRequest`)

File: clients/horizonclient/client.go (hunk 1 above)
Comment (abuiles) [reply clients/horizonclient/client.go]: good catch!

File: clients/horizonclient/main.go (hunk 3 above)
Comment (abuiles) [reply clients/horizonclient/main.go]: @ire-and-curses I agree with you, initially I thought about following a similar approach for `EffectRequest`, but it seems to me that we are trying to do too much with those structs.

That said though, we could be consistent and try to follow a similar approach. Basically extend the `AccountRequest` struct to also take `ForSigner` and `ForAsset`. When called through `client.Accounts` they will be the only fields we look at + Page params, ignoring the rest. 

I personally don't  like that interface, but it is consistent with how the other `*Request` like structs behave.

Comment (leighmcculloch) [reply clients/horizonclient/main.go]: I think this is okay given that we can't make it more consistent with other functions without it being backwards compatible, and the alternatives of smooshing the two types together into a single shared types would be worse and could result in bad human errors. The compiler will make sure the caller uses the right type so whilst there might be some confusion with similar named types any errors arising from that should be prevented.

**Pull request merged**